Fix possible permanent "Cannot schedule a task" error

azat opened the pull request:
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix possible permanent "Cannot schedule a task" error (due to unhandled exception in `ParallelAggregatingBlockInputStream::Handler::onFinish/onFinishThread`)

Refs: #6833

azat committed:
Call onException if ParallelInputsHandler::onFinish* throws

azat commented:
> 00974_query_profiler | FAIL

Fails on upstream/master too

alexey-milovidov reviewed (approved):
Ok but I don't understand what is the fix.

alexey-milovidov commented:
No logic in ThreadPool was changed. It should work with unhandled exceptions regardless to the change.

azat commented:
>No logic in ThreadPool was changed. It should work with unhandled exceptions regardless to the change

In case of exception it will shutdown it, and in some cases the global pool can be terminated, which will lead to "cannot schedule task" permanently for every new connection/query

https:github.com/ClickHouse/ClickHouse/pull/9154/commits/d83f24986ca67669238372710935a2cfea41ac6f has some details, if this is not enough, ping me, I can try to get back with reproducer

alexey-milovidov commented:
> In case of exception it will shutdown it, and in some cases the global pool can be terminated, which will lead to "cannot schedule task" permanently for every new connection/query

We don't use GlobalThreadPool directly, only via ThreadFromGlobalPool.

alexey-milovidov commented:
The proper fix will be:
- don't throw from the job function of ThreadFromGlobalPool.
https://clickhouse-test-reports.s3.yandex.net/codebrowser/html_report/ClickHouse/dbms/src/Common/ThreadPool.h.html#164

azat commented:
>don't throw from the job function of ThreadFromGlobalPool.

Ok.

Does `onFinish*` should be wrapped with `try`/`catch`/`onException` anyway? (looks like it still worth adding)

azat commented on dbms/src/DataStreams/ParallelInputsProcessor.h:
```diff
@@ -217,7 +217,14 @@ class ParallelInputsProcessor
+        }
+        catch (...)
+        {
+            handler.onException(exception, thread_num);
```
FWIW `exception` is wrong here anyway (should be `std::current_exception()`)

azat commented:
>don't throw from the job function of ThreadFromGlobalPool.

So the exception need to be re-thrown somehow, this can be done from the `join()`, but in this case the following should be done:
- modify each `join()` that called in destructor -- not a problem
- `ThreadPoolImpl` will not known that exception occurred until destructor (`finalize`)

I'm suggesting to not shutdown the loop on exceptions, for now this can done only if this is the global pool, but I would prefer not to add this condition, since this will make it less readable and the exception will be thrown from the `wait` anyway, but I can miss something, @alexey-milovidov ?

azat committed:
Do not shutdown global thread pool on exception

Otherwise GlobalThreadPool can be terminated (for example due to an
exception from the ParallelInputsHandler::onFinish/onFinishThread, from
ParallelAggregatingBlockInputStream::Handler::onFinish/onFinishThread,
since writeToTemporaryFile() can definitelly throw) and the server will
not accept new connections (or/and execute queries) anymore.

Here is possible stacktrace (it is a bit inaccurate, due to
optimizations I guess, and it had been obtained with the
DB::tryLogCurrentException() in the catch block of the
ThreadPoolImpl::worker()):

    2020.02.16 22:30:40.415246 [ 45909 ] {} <Error> ThreadPool: Unhandled exception in the ThreadPool(10000,1000,10000) the loop will be shutted down: Code: 241, e.displayText() = DB::Exception: Memory limit (total) exceeded: would use 279.40 GiB (attempt to allocate chunk of 4205536 bytes), maximum: 279.40 GiB, Stack trace (when copying this message, always include the lines below):

    1.  Common/Exception.cpp:35: DB::Exception::Exception(...)
    ...
    6.  Common/Allocator.h:102: void DB::PODArrayBase<8ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::reserve<>(unsigned long) (.part.0)
    7.  Interpreters/Aggregator.cpp:1040: void DB::Aggregator::writeToTemporaryFileImpl<...>(...)
    8.  Interpreters/Aggregator.cpp:719: DB::Aggregator::writeToTemporaryFile(...)
    9.  include/memory:4206: DB::Aggregator::writeToTemporaryFile(...)
    10. DataStreams/ParallelInputsProcessor.h:223: DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>::thread(...)

Refs: https://github.com/ClickHouse/ClickHouse/issues/6833#issuecomment-579221732
(Reference to particular comment, since I'm not sure about the initial issue)

azat commented:
>I'm suggesting to not shutdown the loop on exceptions, for now this can done only if this is the global pool, but I would prefer not to add this condition

Decided to make it for global pool only (at least for now)

filimonov commented:
@azat Is it possible to create a test case for that?

azat commented:
> Is it possible to create a test case for that?

It is easy to cover this with unit test, but this is not that interesting since this is the problem of the callers.
It is also possible to reproduce this via some query (i.e. function? test), but it is not that easy, since the exception should be thrown from the merge stage only (but this can be reproduced, by using the partition with exact size)

azat commented:
>It is also possible to reproduce this via some query (i.e. function? test), but it is not that easy, since the exception should be thrown from the merge stage only (but this can be reproduced, by using the partition with exact size)

So I tried to do this just for fun, but because of compression (for temporary files in aggregator) this approach cannot be used.

>What is happening with 00974_query_profiler ?

I have no idea, but after rebase it does not fails anymore (maybe some issue that has been fixed in master already? anyway logs wasn't verbose enough to investigate this)

alexey-milovidov commented:
@azat query profiler test was temporary broken in master a few weeks ago.
See https://github.com/ClickHouse/ClickHouse/pull/9472#issuecomment-593719032

alexey-milovidov commented:
The original issue
> ParallelAggregatingBlockInputStream::Handler::onFinish/onFinishThread

should not affect versions 20.3+, because `experimental_use_processors` is enabled.

azat commented:
>should not affect versions 20.3+, because experimental_use_processors is enabled.

It was found on 20.2.1.2337-23 (even though AFAIK there wasn't official release), and AFAICS the processors has been enabled:

```
$ git show v20.2.1.2337-testing:dbms/src/Core/Settings.h | fgrep processors
    M(SettingBool, experimental_use_processors, true, "Use processors pipeline.", 0) \
```

alexey-milovidov commented:
@azat But how it's related to `ParallelInputsProcessor`, that (despite it's name) has no relationship to processors?

azat commented:
>@azat But how it's related to ParallelInputsProcessor, that (despite it's name) has no relationship to processors?

Hm, query that triggers it had been initiated by the Distributed engine, and I guess [it was possible](https://github.com/ClickHouse/ClickHouse/pull/8929/files#diff-674a9476cad518ac771ee9d89fa4f282R92) in pre #8929 (v20.2.1.2425-testing-181-g1e8389eceb)

alexey-milovidov commented:
Sorry, this PR introduced race condition:
https://clickhouse-test-reports.s3.yandex.net/11301/8a4c2380ddd7a1e316f4d3d1eb5d178bccabec1f/stress_test_(thread)/stderr.log

(I will fix it)

alexey-milovidov commented:
@azat we suspect that this issue is not fixed: https://github.com/ClickHouse/ClickHouse/issues/33712
